[content-visibility] ASSERTION FAILED: !renderer.needsLayout() causing imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-video.html to constantly crash

obyknovenius opened the pull request:
#### 836cb999ce8c37075ae7c05ef485328a4b96e73c
<pre>
[content-visibility] ASSERTION FAILED: !renderer.needsLayout() causing imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-video.html to constantly crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=262157">https://bugs.webkit.org/show_bug.cgi?id=262157</a>

Reviewed by Darin Adler.

Normally, when we do `RenderElement::layout()`, we first lay out its
children and then clear the `needsLayout` flag. When assertions are
enabled, in `RenderObject::clearNeedsLayout()` we check that
block-positioned children don&apos;t need layout.

But if the element is skipped for layout, we call
`RenderElement::clearNeedsLayoutForSkippedContent()` instead.

As the first step, this method calls `RenderObject::clearNeedsLayout()`
on every descendent. Note that it won&apos;t lay out the descendant&apos;s
children but will still check that the block-positioned children of the
descendent don&apos;t need layout, which might be false and cause the
assertion to fail.

This patch implements `descendantsOfTypePostOrder&lt;T&gt;()` function which
is then used to guarantee that `RenderElement`&apos;s needs-layout flag is
always cleared after its children.

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebCore/rendering/RenderDescendantIterator.h:
(WebCore::RenderDescendantPostOrderIterator&lt;T&gt;::RenderDescendantPostOrderIterator):
(WebCore::RenderDescendantPostOrderIterator&lt;T&gt;::operator):
(WebCore::RenderDescendantPostOrderConstIterator&lt;T&gt;::RenderDescendantPostOrderConstIterator):
(WebCore::RenderDescendantPostOrderConstIterator&lt;T&gt;::operator):
(WebCore::RenderDescendantPostOrderIteratorAdapter&lt;T&gt;::RenderDescendantPostOrderIteratorAdapter):
(WebCore::RenderDescendantPostOrderIteratorAdapter&lt;T&gt;::begin):
(WebCore::RenderDescendantPostOrderIteratorAdapter&lt;T&gt;::end):
(WebCore::RenderDescendantPostOrderIteratorAdapter&lt;T&gt;::at):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter&lt;T&gt;::RenderDescendantPostOrderConstIteratorAdapter):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter&lt;T&gt;::begin const):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter&lt;T&gt;::end const):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter&lt;T&gt;::at const):
(WebCore::descendantsOfTypePostOrder):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::clearNeedsLayoutForSkippedContent):
* Source/WebCore/rendering/RenderIterator.h:
(WebCore::RenderPostOrderIterator::operator bool const):
(WebCore::RenderPostOrderConstIterator::operator bool const):
(WebCore::RenderObjectPostOrderTraversal::next):
(WebCore::RenderPostOrderTraversal::firstWithin):
(WebCore::RenderPostOrderTraversal::next):
(WebCore::RenderPostOrderIterator&lt;T&gt;::RenderPostOrderIterator):
(WebCore::RenderPostOrderIterator&lt;T&gt;::traverseNext):
(WebCore::RenderPostOrderIterator&lt;T&gt;::operator):
(WebCore::= const):
(WebCore::RenderPostOrderConstIterator&lt;T&gt;::RenderPostOrderConstIterator):
(WebCore::RenderPostOrderConstIterator&lt;T&gt;::traverseNext):
(WebCore::RenderPostOrderConstIterator&lt;T&gt;::operator const):
(WebCore:: const):

Canonical link: <a href="https://commits.webkit.org/277019@main">https://commits.webkit.org/277019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3918e6f1d1d60c5991ea9aee705e35f333991a39

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46424 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25570 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49023 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49099 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48731 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29932 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23023 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47002 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22609 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40029 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19128 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41134 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4469 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41503 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50952 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17870 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10274 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22419 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2487,7 +2487,7 @@ bool RenderElement::hasEligibleContainmentForSizeQuery() const
 {
     for (CheckedRef descendant : descendantsOfType<RenderObject>(*this))
-        descendant->clearNeedsLayout(EverHadSkippedContentLayout::No);
+        descendant->clearNeedsLayoutForSkippedContent();
```
This is not good. It will be highly inefficient if one of the descendants is itself a `RenderElement`. We will loop over the same `RenderObject` one time for each `RenderElement` enclosing it. This should be left as-is.

rwlbuis replied:
Besides Darin's comment, is the ASSERT the same as described in the bug? I landed some PRs that were expected to avoid this dirty renderer problem. If so, which RenderElement is not laid out? Extra points if you can show a render tree dump (showRenderTreeForThis).

obyknovenius replied:
> Besides Darin's comment, is the ASSERT the same as described in the bug?

Yes, it is:
```
ASSERTION FAILED: !renderer.needsLayout()
/Users/vitaly/Developer/WebKit/Source/WebCore/rendering/RenderBlock.cpp(3174) : void WebCore::RenderBlock::checkPositionedObjectsNeedLayout()
com.apple.WebKit.WebContent.Development terminated (pid 39643) for reason: crash
```

rwlbuis replied:
Do you know which renderer has dirty layout? 
I think my fixes were for out-of-flow positioned content, if the test does not contain out-of-flow content, and it does not look like it has, then that is an obvious reason my previous fixes would not work here.

obyknovenius replied:
> Extra points if you can show a render tree dump (showRenderTreeForThis).

```
(B)lock/(I)nline/I(N)line-block, (A)bsolute/Fi(X)ed/(R)elative/Stic(K)y, (F)loating, (O)verflow clip, Anon(Y)mous, (G)enerated, has(L)ayer, hasLayer(S)crollableArea, (C)omposited, Content-visibility:(H)idden/(A)uto, (S)kipped content, (+)Dirty style, (+)Dirty layout
B---YGLSC-- --  RenderView at (0,0) size 800x600 renderer (0x1150002c0) layout box (0x0)
B-----LSC-- --    HTML RenderBlock at (0,0) size 800x222 renderer (0x115001700) layout box (0x0) node (0x115001050)
B---------- --      BODY RenderBody at (8,8) size 784x206 renderer (0x115001a20) layout box (0x11500ea20) node (0x115001250)
---------- --        line at (0.00,0.00) size (784.00x206.00) baseline (202.00) enclosing top (0.00) bottom (202.00)
---------- --          Root inline box at (0.00,188.00) size (202.00x18.00)
---------- --          Run(s):
---------- --            Atomic box at (0.00,0.00) size 202.00x202.00 renderer->(0x11500e810)
I-----L-CH- --        VIDEO RenderVideo at (0,0) size 202x202 renderer (0x11500e810) layout box (0x11500ead0) node (0x115005550)
BR----L---S -+          DIV RenderBlock at (1,1) size 200x0 renderer (0x115003630) layout box (0x0) node (0x115006350) layout->[normal child][positioned child]
BA----L---S -+            DIV RenderBlock at (0,0) size 200x200 renderer (0x1150038d0) layout box (0x0) node (0x115006530) layout->[self][normal child][positioned child]
BA----L---S --              BUTTON RenderFlexibleBox at (100,100) size 60x60 renderer (0x115011af0) layout box (0x0) node (0x115006810)
BA----L---S -+*               DIV RenderBlock at (0,0) size 60x60 renderer (0x115011ea0) layout box (0x0) node (0x11500cd20) layout->[self][normal child]
BA----L---S -+                  DIV RenderBlock at (0,0) size 60x60 renderer (0x115012140) layout box (0x0) node (0x11500cda0) layout->[self]
BA----L---S -+                  DIV RenderBlock at (0,0) size 60x60 renderer (0x1150123e0) layout box (0x0) node (0x11500ce20) layout->[self]
BA----L---S -+                PICTURE RenderBlock at (3,0) size 60x60 renderer (0x115012680) layout box (0x0) node (0x1150069d0) layout->[self]
I---------- --        #text RenderText renderer (0x115011550) layout box (0x115011620) node (0x115011310) length->(3) "\n\n\n"
I---------- --        #text RenderText renderer (0x115011800) layout box (0x1150116c0) node (0x1150113f0) length->(2) "\n\n"
```

obyknovenius replied:
> This is not good. It will be highly inefficient if one of the descendants is itself a `RenderElement`. We will loop over the same `RenderObject` one time for each `RenderElement` enclosing it. This should be left as-is.

You're right. I've missed the fact that we're traversing all descendants. We should use `childrenOfType` here.

rwlbuis replied:
Thanks! I lack time to investigate, maybe end of the week I have some time. If you have some cycles before that, I would point out that RenderVideo is inline rather than block, and also not a RenderBox, and I wonder if the content-visibility logic does enough clearing of the layout bit in this case.

darinadler reviewed (changes requested):
I don’t understand the point of this change. The new code does the same thing as the old, clearing all the needs-layout flags, but does it less efficiently with a recursive algorithm and virtual function. How does that fix a bug?

darinadler commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2483,10 +2483,22 @@ bool RenderElement::hasEligibleContainmentForSizeQuery() const
+            function(child);
+    }
+    function(parent);
+};
```
It’s OK to do it in this functional style, although ideally if this comes up a lot we will want to add `descentsOfTypePostOrder` to `RenderDescendantIterator.h`.

But we want a non-recursive algorithm. Here’s one cut at that:
```
static RenderObject& firstInPostOrder(RenderObject& object)
{
    if (auto* firstLeafChild = next->firstLeafChild())
        return *firstLeafChild;
    return object;
}

static RenderObject* nextInPostOrder(RenderObject& object)
{
    if (auto* next = object->nextSibling())
        return &firstInPostOrder(*next);
    return object->parent();
}

static void forEachDescendantInPostOrder(void (*function)(RenderObject&), RenderElement& parent)
{
    for (auto* next = &firstInPostOrder(parent); next != &parent; next = nextInPostOrder(*next)) {
        function(*next);
    }
}
```

obyknovenius replied:
@darinadler  I've implemented `descendantsOfTypePostOrder` using your code snippet as a base. Thank you for the guidance!

darinadler commented on Source/WebCore/rendering/RenderIterator.h:
```diff
@@ -210,6 +275,28 @@ inline T* next(U& current, const RenderObject* stayWithin)
+    auto* descendant = RenderObjectPostOrderTraversal::firstLeafChild(current);
+    while (descendant && !isRendererOfType<T>(*descendant))
+        descendant = RenderObjectPostOrderTraversal::next(*descendant, &current);
+    return static_cast<T*>(descendant);
```
I think we want to return `nullptr` here if `descendant == &current` and because `RenderObjectPostOrderTraversal::firstLeafChild` is slightly imprecisely named it’s not obvious. The symptom would be that for an object without children we would iterate the ancestor itself when calling `descendantsOfTypePostOrder` rather than iterating nothing. Harmless for `RenderElement::clearNeedsLayoutForSkippedContent` but not correct. I suggest fixing this by moving the null checking from `RenderObjectPostOrderTraversal::firstLeafChild` into `RenderObjectPostOrderTraversal::next`.

obyknovenius replied:
You're right. I've fixed it the way you suggested. I also noticed that when an object has children it's iterated as a last child which is wrong. I've added another check to fix it. Also, I've made `RenderObjectPostOrderTraversal`s inlines non-templated since it's not really necessary.

darinadler commented on Source/WebCore/rendering/RenderIterator.h:
```diff
@@ -140,8 +181,39 @@ inline RenderObject* nextSkippingChildren(RenderObject& current, const RenderObj
+
+namespace RenderObjectPostOrderTraversal {
+
+inline RenderObject* firstLeafChild(RenderObject& object)
```
We don't need this function any more. Can just call `RenderObject::firstLeafChild` directly.

obyknovenius replied:
Done.

darinadler commented on Source/WebCore/rendering/RenderIterator.h:
```diff
@@ -140,8 +181,39 @@ inline RenderObject* nextSkippingChildren(RenderObject& current, const RenderObj
+    return object.firstLeafChild();
+}
+
+inline RenderObject* nextAncestor(RenderObject& current, const RenderObject* stayWithin)
```
Not sure I would have added this helper function. I would have just put this logic into the `next` function.

obyknovenius replied:
Done.

darinadler commented on Source/WebCore/rendering/RenderIterator.h:
```diff
@@ -140,8 +181,29 @@ inline RenderObject* nextSkippingChildren(RenderObject& current, const RenderObj
+
+inline RenderObject* next(RenderObject& current, const RenderObject* stayWithin)
+{
+    if (&current == stayWithin)
```
I just realized this check will never be true; it’s dead code. So we could remove it. The one you added after calling `parent` is the actual check needed.

obyknovenius replied:
Done.

darinadler reviewed (approved):
Yes I agree this is OK to land as is but I found one more loose end.

obyknovenius committed:
[content-visibility] ASSERTION FAILED: !renderer.needsLayout() causing imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-video.html to constantly crash
https://bugs.webkit.org/show_bug.cgi?id=262157

Reviewed by Darin Adler.

Normally, when we do `RenderElement::layout()`, we first lay out its
children and then clear the `needsLayout` flag. When assertions are
enabled, in `RenderObject::clearNeedsLayout()` we check that
block-positioned children don't need layout.

But if the element is skipped for layout, we call
`RenderElement::clearNeedsLayoutForSkippedContent()` instead.

As the first step, this method calls `RenderObject::clearNeedsLayout()`
on every descendent. Note that it won't lay out the descendant's
children but will still check that the block-positioned children of the
descendent don't need layout, which might be false and cause the
assertion to fail.

This patch implements `descendantsOfTypePostOrder<T>()` function which
is then used to guarantee that `RenderElement`'s needs-layout flag is
always cleared after its children.

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebCore/rendering/RenderDescendantIterator.h:
(WebCore::RenderDescendantPostOrderIterator<T>::RenderDescendantPostOrderIterator):
(WebCore::RenderDescendantPostOrderIterator<T>::operator):
(WebCore::RenderDescendantPostOrderConstIterator<T>::RenderDescendantPostOrderConstIterator):
(WebCore::RenderDescendantPostOrderConstIterator<T>::operator):
(WebCore::RenderDescendantPostOrderIteratorAdapter<T>::RenderDescendantPostOrderIteratorAdapter):
(WebCore::RenderDescendantPostOrderIteratorAdapter<T>::begin):
(WebCore::RenderDescendantPostOrderIteratorAdapter<T>::end):
(WebCore::RenderDescendantPostOrderIteratorAdapter<T>::at):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter<T>::RenderDescendantPostOrderConstIteratorAdapter):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter<T>::begin const):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter<T>::end const):
(WebCore::RenderDescendantPostOrderConstIteratorAdapter<T>::at const):
(WebCore::descendantsOfTypePostOrder):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::clearNeedsLayoutForSkippedContent):
* Source/WebCore/rendering/RenderIterator.h:
(WebCore::RenderPostOrderIterator::operator bool const):
(WebCore::RenderPostOrderConstIterator::operator bool const):
(WebCore::RenderObjectPostOrderTraversal::next):
(WebCore::RenderPostOrderTraversal::firstWithin):
(WebCore::RenderPostOrderTraversal::next):
(WebCore::RenderPostOrderIterator<T>::RenderPostOrderIterator):
(WebCore::RenderPostOrderIterator<T>::traverseNext):
(WebCore::RenderPostOrderIterator<T>::operator):
(WebCore::= const):
(WebCore::RenderPostOrderConstIterator<T>::RenderPostOrderConstIterator):
(WebCore::RenderPostOrderConstIterator<T>::traverseNext):
(WebCore::RenderPostOrderConstIterator<T>::operator const):
(WebCore:: const):

Canonical link: https://commits.webkit.org/277019@main
